mirrord/kube/src/api/container/job.rs

```diff
@@ -116,6 +186,37 @@ where
     })
 }
 
+/// Tries finding mirrord-agent's container in `status.container_statuses` and returns a string
+/// representing it's state.
+fn find_agent_container_state(status: &Option<PodStatus>) -> Option<&'static str> {
+    status
+        .as_ref()
+        .and_then(|status| status.container_statuses.clone())
+        .and_then(|statuses| {
+            statuses
+                .iter()
+                .find(|container| container.name == "mirrord-agent")
+                .cloned()
+        })
+        .and_then(|status| status.state)
+        .and_then(|state| {
+            // `ContainerState` is essentially an enum pretending to be a struct, so we have to this
+            // little dance, which really wants to be a pattern match.
+            if state.running.is_some() {
+                Some("running")
+            } else if state.terminated.is_some() {
+                Some("terminated")
+            } else if state.waiting.is_some() {
+                Some("waiting")
+            } else {
+                // This should be unreachable - the docs say that if the state is unknown then the
+                // active state will be 'waiting', but it doesn't sound nice to
+                // crash if another state ever gets added.
+                None
+            }
+        })
+}
```

Comment:
```suggestion
/// Tries finding mirrord-agent's container in `status.container_statuses` and returns a string
/// representing it's state.
fn find_agent_container_state(status: &Option<PodStatus>) -> String {
    status
        .as_ref()
        .into_iter()
        .flat_map(|status| status.container_statuses.as_deref().unwrap_or_default())
        .find(|status| status.name == "mirrord-agent")
        .and_then(|status| status.state.as_ref())
        .and_then(|state| {
            if state.running.is_some() {
                Some("running".into())
            } else if let Some(terminated) = &state.terminated {
                Some(format!(
                    "terminated ({})",
                    terminated.reason.as_deref().unwrap_or("no reason found")
                ))
            } else {
                state.waiting.as_ref().map(|waiting| {
                    format!(
                        "waiting ({})",
                        waiting.reason.as_deref().unwrap_or("no reason found")
                    )
                })
            }
        })
        .unwrap_or_else(|| "not found".into())
}
```
Less cloning, and reason included ^^